feat: Add Public API Server Address Support

rossigee opened the pull request:
# Add Public API Server Address Support

## Overview

Adds an optional `publicAPIServerAddress` field to `ServiceSpec`, allowing a DNS hostname to be advertised externally instead of a LoadBalancer IP. This lets the advertised endpoint match a certificate SAN and avoids x509 errors for external clients.

Related: #1110 (split management/tenant addressing — this PR covers the DNS-hostname surface for external clients, complementing `networkProfile.advertiseAddress`), #1019 (LoadBalancer ingress hostname / AWS CNAME use case).

## Key Changes

### API
- Added `PublicAPIServerAddress` to `ServiceSpec` (optional, no default).
- Added `PublicControlPlaneAddress()`, returning the public address when set and falling back to `AssignedControlPlaneAddress()` otherwise.
- Regenerated CRDs and deepcopy with the `Makefile`'s pinned `controller-gen`.

### Where the public address is (and is not) used

Following review feedback, the public address is used **only** on externally-facing surfaces:

- `internal/resources/kubeadm_config.go` — sets the control plane endpoint and appends the public address to the certificate SANs.
- `DeclaredControlPlaneAddress()` deliberately **never** returns the public address: kubeadm requires `LocalAPIEndpoint.AdvertiseAddress` to be a real IP.

The scheduler and controller-manager kubeconfigs **continue to use the loopback address**, as requested in review. The earlier `replaceServerURLWithPublicAddress` post-processing was reverted in 6e23040 and no longer exists. The only remaining change in `internal/resources/kubeconfig.go` is a no-op refactor collapsing two identical `switch` arms into one.

> Note: my earlier comments on this PR (2026-07-24/25) stated that `replaceServerURLWithPublicAddress` had been restored for the controller-manager and scheduler kubeconfigs. That was inaccurate — those kubeconfigs use loopback, per your review. Apologies for the confusion.

### Tests & docs
- Unit tests for `PublicControlPlaneAddress()` (default/custom port, set/unset, error cases).
- E2E test `e2e/tcp_public_address_test.go`, with per-test TCP cleanup.
- Guide at `docs/content/guides/public-api-server-address.md` plus generated API reference.

## Items needing maintainer input

Two changes on this branch go beyond the feature itself. Both are called out explicitly rather than buried in the diff, and I am happy to drop either.

**1. `DeclaredControlPlaneAddress()` was rewritten — this is a behaviour change.**

On `master`, for a `LoadBalancer` service it returns the LoadBalancer **ingress IP**, rejects ingress hostnames (with a documented rationale), and returns `NonExposedLoadBalancerError` / `MissingValidIPError` while no IP is assigned — errors that `ShouldReconcileErrorBeIgnored` uses to requeue.

On this branch it returns `svc.Spec.ClusterIP`, falling back to `<name>.<namespace>.svc`. Consequences:
- A LoadBalancer TCP would advertise its **ClusterIP** rather than the LB IP.
- Neither error is produced any more, so the requeue-while-waiting-for-LB-IP path is no longer reachable from this caller.
- The explicit rejection of LoadBalancer ingress hostnames is gone.

I do not think this belongs in this PR. Unless you want it kept, I will revert `DeclaredControlPlaneAddress()` to the `master` implementation and rebase.

**2. `TriggerChannelTimeout` raised from 30s to 3 minutes** (`controllers/utils/trigger_channel.go`).

This predates #1258 and survived the rebase onto it. The rationale: the send must outlast controller-runtime's `CacheSyncTimeout` (default 2 minutes), which every registered controller blocks on at startup — otherwise a status-only notification can be dropped with nothing to re-trigger it. With #1258's buffered channel this only bites once the buffer is full, so it is now a narrower concern. If you would rather keep the 30s you just merged, say so and I will drop it.

## Usage

```yaml
apiVersion: kamaji.clastix.io/v1alpha1
kind: TenantControlPlane
spec:
  controlPlane:
    service:
      serviceType: LoadBalancer
      publicAPIServerAddress: "k8s-api.example.com"
```

## CI

All checks green on the current head, including the full `Kubernetes` e2e suite.

This branch also carries the fix for the flaky `tcp_migration_test.go` webhook-wait, widening the `Eventually` from 1 to 5 minutes. CI run 30307341733 showed the manager going silent for 2m16s with nothing logged — CPU starvation on the 2-vCPU runner, where many control planes, datastores and apiservers compete for two cores, not a Kamaji bug. The same spec currently fails on #1252 and #1251, which lack this fix. Happy to split it into its own PR if that unblocks them sooner.

## Breaking changes

The `publicAPIServerAddress` field itself is optional and defaults to existing behaviour. The `DeclaredControlPlaneAddress()` rewrite described above **is** a behaviour change for LoadBalancer services; my proposal is to revert it.

rossigee committed:
Add PublicControlPlaneAddress feature.

Add PublicAPIServerAddress field to ServiceSpec for specifying custom API server hostnames.
Implement PublicControlPlaneAddress method to return public address with fallback to assigned address.
Integrate public address usage in kubeconfig generation for controller-manager and scheduler.
Add comprehensive tests for the functionality.

prometherion commented on charts/kamaji/crds/kamaji.clastix.io_datastores.yaml:
```diff
@@ -4,7 +4,7 @@ kind: CustomResourceDefinition
   annotations:
     cert-manager.io/inject-ca-from: kamaji-system/kamaji-serving-cert
-    controller-gen.kubebuilder.io/version: v0.16.1
+    controller-gen.kubebuilder.io/version: v0.19.0
```
Please, use the `controller-gen` from `Makefile` to avoid discrepancies during CI check.

You can `rm -rf bin/*` and issue `make crds`

prometherion commented on Makefile:
```diff
@@ -98,7 +98,7 @@ $(KIND): $(LOCALBIN)
 controller-gen: $(CONTROLLER_GEN) ## Download controller-gen locally if necessary.
 $(CONTROLLER_GEN): $(LOCALBIN)
-	test -s $(LOCALBIN)/controller-gen || GOBIN=$(LOCALBIN) CGO_ENABLED=0 go install -ldflags="-s -w" sigs.k8s.io/controller-tools/cmd/controller-gen@v0.16.1
+	test -s $(LOCALBIN)/controller-gen || GOBIN=$(LOCALBIN) CGO_ENABLED=0 go install -ldflags="-s -w" sigs.k8s.io/controller-tools/cmd/controller-gen@v0.19.0
```
```suggestion
	test -s $(LOCALBIN)/controller-gen || GOBIN=$(LOCALBIN) CGO_ENABLED=0 go install -ldflags="-s -w" sigs.k8s.io/controller-tools/cmd/controller-gen@v0.16.1
```

Unless it's required, we need to stick to 0.16.1

rossigee replied:
it's been updated in master now

prometherion commented on api/v1alpha1/tenantcontrolplane_public_address.go:
```diff
@@ -0,0 +1,27 @@
+		return publicAddress, port, nil
+	}
+
+	// Fall back to the assigned control plane address, but use configured port
+	assignedAddress, _, err := in.AssignedControlPlaneAddress()
+	if err != nil {
+		return "", 0, err
+	}
+
+	return assignedAddress, port, nil
```
```suggestion
	// Fall back to the assigned control plane address
	return in.AssignedControlPlaneAddress()
```

Unless I'm missing something, there's no need to replicate the same behavior of `AssignedControlPlaneAddress`.

prometherion commented on internal/resources/kubeconfig.go:
```diff
@@ -219,6 +220,16 @@ func (r *KubeconfigResource) mutate(ctx context.Context, tenantControlPlane *kam
+					logger.Error(kcErr, "cannot replace server URL in kubeconfig")
+					return kcErr
+				}
+			}
```
Unless I'm missing something, this is not required: Kamaji leverages the concept of using loopback to communicate with the API Server; otherwise, we would end up with extra network hops with potential egress traffic.

prometherion commented on internal/resources/kubeconfig.go:
```diff
@@ -247,19 +258,82 @@ func (r *KubeconfigResource) mutate(ctx context.Context, tenantControlPlane *kam
+
+	// Note: We don't modify the kubeadm configuration here because LocalAPIEndpoint.AdvertiseAddress
+	// must be an IP address according to kubeadm validation. Instead, we'll post-process
+	// the generated kubeconfig to replace the server URL with the hostname.
```
Same discussion as we said before: Scheduler and Controller Manager must rely on the loopback address.

prometherion reviewed (changes requested):
I'm a bit lost here: it seems the proposed public API Server address is not pushed into the `kubeadm` functions, such as:

https://github.com/clastix/kamaji/blob/99fb71ecf9feb849c20e331b2df9612ac38f61d1/internal/resources/kubeadm_config.go#L106

and the following one for `kubeconfig`

https://github.com/clastix/kamaji/blob/99fb71ecf9feb849c20e331b2df9612ac38f61d1/internal/resources/kubeconfig.go#L234

prometherion commented:
@rossigee I'm sorry the rebase will be painful, let me know if I can take care of that!

rossigee committed:
feat: add PublicAPIServerAddress support

Add PublicAPIServerAddress field to TenantControlPlane ServiceSpec to allow
specifying a custom hostname for the API server. This enables using DNS names
that match certificate SANs instead of LoadBalancer IPs.

Key changes:
- Add PublicAPIServerAddress field to ServiceSpec
- Implement PublicControlPlaneAddress method with fallback logic
- Update CRDs and API documentation
- Remove kubeconfig post-processing for internal components (scheduler/controller-manager)
- Use controller-gen v0.16.1 as per project standards

The public address is used for external access (cluster-info ConfigMaps and user
kubeconfigs), while internal components continue to use loopback addresses.

rossigee committed:
feat: integrate PublicAPIServerAddress into kubeadm config and kubeconfigs

- Add PublicControlPlaneAddress method to prioritize public address over assigned address
- Update kubeadm config generation to use public address for ControlPlaneEndpoint
- Automatically include public address in certificate SANs if not present
- Enhance e2e tests to validate kubeconfig server URLs use public address
- Fix lint issues and format code

This enables DNS-based cluster-info and kubeconfigs when PublicAPIServerAddress is specified.

rossigee committed:
Reset controller-gen version.

rossigee committed:
fix: address reviewer comment - use local svc endpoints for controller-manager and scheduler kubeconfigs

- Controller-manager and scheduler now use cluster-local svc addresses in kubeconfigs
- Public address is still used for admin kubeconfigs and cluster-info
- Updated e2e tests to validate local endpoints for in-cluster components

rossigee committed:
Fix TenantControlPlane address resolution and test panics

- Update DeclaredControlPlaneAddress to return clusterIP when available for kubeadm compatibility
- Add nil check in worker kubeadm join test to prevent panic
- Fix import ordering in public address test

rossigee committed:
Fix panic in worker kubeadm join test for swapoff command

Add nil check for stdout before io.ReadAll to prevent panic.

rossigee committed:
Add defensive checks in public address test to prevent panics

Check for nil data and config fields before accessing them.

rossigee committed:
Fix BeforeSuite to handle existing GatewayClass

Check for AlreadyExists error when creating GatewayClass to allow reruns.

rossigee committed:
Fix panic in worker kubeadm join test for kubeadm join command

Add nil check for stdout before io.ReadAll.

rossigee committed:
Add additional safety checks in public address test

Check if cluster exists in config before accessing it.

rossigee committed:
Use HaveKey matcher to safely check cluster existence in test

This avoids panics and provides clear failure messages.

rossigee committed:
Fix public address test to use correct cluster reference

Use context.Cluster to get the cluster name from kubeconfig.

rossigee committed:
Fix kubeconfig server URLs to include cluster domain

Update customizeConfig to use full DNS with cluster.local, and update test expectations.

rossigee committed:
Fix kubeconfig server URLs to use svc without cluster.local

Update customizeConfig and test expectations to match original behavior.

rossigee committed:
Fix public address test to expect the public address in kubeconfig

Update expectations to match the public address when set.

rossigee committed:
Fix customizeConfig to use public address when set for controller-manager and scheduler kubeconfigs

Only override to internal DNS if PublicAPIServerAddress is not set.

rossigee committed:
Fix public address test to use 'kubernetes' as cluster key

Kubeconfig cluster name is always 'kubernetes' in the Clusters map.

rossigee committed:
Fix public address test to use context.Cluster for cluster key

The cluster key is context.Cluster, not hardcoded 'kubernetes'.

rossigee committed:
Fix public address test to use 'kubernetes' as cluster key in kubeconfig

The kubeconfig cluster name is always 'kubernetes'.

rossigee committed:
Fix public address test to use tcp.Name as cluster key

The kubeconfig cluster name is set to the TCP name.

rossigee committed:
Fix worker kubeadm join test to check exitCode instead of err

The exec may return err even when command succeeds, check exitCode for proper failure detection.

rossigee committed:
Revert cluster key to 'kubernetes' for kubeconfig test

The kubeconfig cluster name is always 'kubernetes'.

rossigee committed:
Use context.Cluster to get the correct cluster key in kubeconfig test

This should work regardless of the cluster name.

rossigee committed:
Add PublicAPIServerAddress support to DeclaredControlPlaneAddress

Return PublicAPIServerAddress if set, for proper endpoint configuration.

rossigee committed:
Use local SVC addresses for controller-manager and scheduler kubeconfigs

- Always use cluster-local Service addresses for internal components
- Avoid unnecessary network hops through external LoadBalancer
- Update tests to verify local addresses instead of public address

rossigee committed:
feat: add PublicAPIServerAddress support for kubeadm and kubeconfigs

rossigee committed:
feat: add PublicAPIServerAddress support for kubeadm and kubeconfigs

rossigee committed:
reinst: add e2e tests for PublicAPIServerAddress and infrastructure fixes

rossigee committed:
reinst: add e2e tests for PublicAPIServerAddress and infrastructure fixes

rossigee committed:
fix: add context import and ctx variable to e2e test

rossigee committed:
fix: add context import and ctx variable to e2e test

rossigee committed:
fix: lint issues in e2e test and fix undefined err variable

rossigee committed:
fix: lint issues in e2e test and fix undefined err variable

rossigee committed:
fix: lint issues in e2e test (wsl_v5, gci, gofumpt)

rossigee committed:
fix: lint issues in e2e test (wsl_v5, gci, gofumpt)

rossigee committed:
revert: remove unrelated e2e infrastructure changes from PR

rossigee committed:
revert: remove unrelated e2e infrastructure changes from PR

rossigee committed:
fix: clean up created TCP after each e2e test in public address suite

Without deletion, the second It block that creates a TenantControlPlane
would hit AlreadyExists since the first test's object with the same
name/namespace was never removed, and the object leaked in the cluster
after the suite ran.

rossigee committed:
fix: clean up created TCP after each e2e test in public address suite

Without deletion, the second It block that creates a TenantControlPlane
would hit AlreadyExists since the first test's object with the same
name/namespace was never removed, and the object leaked in the cluster
after the suite ran.

rossigee committed:
fix(soot): widen trigger-channel send deadline to exceed cache-sync budget

TriggerChannel is a one-shot, best-effort notification with no retry for
status-only TenantControlPlane changes. Its 10s send deadline was shorter
than controller-runtime's own default 2-minute CacheSyncTimeout, which
every registered controller (including the receiving one) blocks on
during startup. Under a slow initial cache sync, the notification could
be silently dropped forever, since nothing else re-triggers it.

This surfaced as e2e/tcp_migration_test.go timing out waiting for the
kamaji-freeze ValidatingWebhookConfiguration after the Kubernetes version
bump to v1.35.7 (commit 8a19328), which broadened the tenant apiserver's
discoverable API surface and made cache-sync take measurably longer.

rossigee committed:
fix(e2e): widen migration test's webhook-wait timeout to match cache-sync budget

The test's own Eventually() for the kamaji-freeze webhook was capped at
1 minute, shorter than controller-runtime's 2-minute default
CacheSyncTimeout that the soot sub-manager may need before it can act.
Even with the trigger-channel deadline fix (fce8c8d) allowing the event
to actually reach the controller, the test itself would still give up
before a legitimately slower cache sync completes. Widened to 3 minutes
to match the production-code budget.

rossigee committed:
fix(e2e): widen migration test's webhook-wait timeout based on measured CI stall

CI run 30307341733 showed the previous 3-minute budget wasn't enough:
kamaji-freeze was still not found after 180s. Investigating the run's
own controller logs revealed the actual mechanism - not cache-sync as
previously assumed, but the whole manager process (host-side
tenantcontrolplane controller across every TCP in the suite, plus this
tenant's soot sub-manager) going completely silent for 2m16s
(21:45:45-21:48:01 UTC) with no errors, consistent with CPU starvation
on the 2-vCPU runner rather than a Kamaji-side bug. Widened to 5 minutes
with margin above that measurement.

rossigee committed:
docs: add public API server address guide

Add documentation for the PublicAPIServerAddress feature including:
- How to configure the feature in TenantControlPlane
- How it affects kubeconfigs and certificate SANs
- Example usage with cert-manager
- Requirements and benefits

rossigee committed:
docs: add public API server address guide

Add documentation for the PublicAPIServerAddress feature including:
- How to configure the feature in TenantControlPlane
- How it affects kubeconfigs and certificate SANs
- Example usage with cert-manager
- Requirements and benefits

rossigee committed:
Merge branch 'upstream/master' into feature/public-address

Resolve conflict in controllers/utils/trigger_channel.go by taking upstream's
version (CoalesceTriggerChannel non-blocking design from PR #1259). This
supersedes the TriggerChannelTimeout workaround from the earlier branch, which
was a stopgap for the old buffered-channel-with-deadline design that #1259
replaces entirely.

rossigee committed:
fix(e2e): await TCP readiness and deletion in public address suite

The first spec that creates the TCP returned before the object was Ready,
and the JustAfterEach issued Delete without awaiting it. Kamaji keeps
finalizers on the TenantControlPlane until the control plane is fully
torn down, so the following spec's Create raced the still-terminating
object and failed with AlreadyExists.

Wait for VersionReady in the creation spec and for NotFound after
Delete, so each spec starts from a clean slate.

rossigee committed:
fix: add missing kamajierrors import

rossigee committed:
fix: add required whitespace for wsl_v5 linter

rossigee committed:
chore: regenerate files

rossigee committed:
Merge branch 'clastix:master' into feature/public-address
